### ISSUE TITLE
Telnet build failures with Xcode 10.

### ISSUE THREAD

**Issue by akhansen** [fink/fink-distributions#368]
There are a couple  of issues here.

- When using `--build-as-nobody` the build fails as follows:
`2019-03-13 21:34:23.251 xcodebuild[81967:974632]  iOSSimulator: [SimServiceContext defaultDeviceSetWithError:] returned nil (Error Domain=com.apple.CoreSimulator.SimError Code=159 "Failed to initialize simulator device set." UserInfo={NSLocalizedFailureReason=Failed to subscribe to notifications from CoreSimulatorService., NSLocalizedDescription=Failed to initialize simulator device set., NSUnderlyingError=0x7f96a12216e0 {Error Domain=com.apple.CoreSimulator.SimError Code=159 "Unable to determine SimDeviceSet, set_path=/var/empty/Library/Developer/CoreSimulator/Devices" UserInfo={NSLocalizedDescription=Unable to determine SimDeviceSet, set_path=/var/empty/Library/Developer/CoreSimulator/Devices, NSUnderlyingError=0x7f96a1221640 {Error Domain=NSPOSIXErrorDomain Code=12 "Cannot allocate memory" UserInfo={NSLocalizedFailureReason=Allocation or initialization failed., NSLocalizedDescription=Failed to initialize simulator device set.}}}}}). Simulator device support disabled.`
`2019-03-13 21:34:24.008 xcodebuild[81967:974627] [MT] IDELogStore: Failed to open log store at /var/empty/Library/Developer/Xcode/DerivedData/libtelnet-eparumqhzcujzafrpztuxqtzgfao/Logs/Build`
`2019-03-13 21:34:24.008 xcodebuild[81967:974627] [MT] IDELogStore: Failed to open Build log store: Error Domain=IDEFoundationErrorDomain Code=1 "Unable to create log store directory at '/var/empty/Library/Developer/Xcode/DerivedData/libtelnet-eparumqhzcujzafrpztuxqtzgfao/Logs/Build': (513) You don’t have permission to save the file “Build” in the folder “Logs”." UserInfo={NSLocalizedDescription=Unable to create log store directory at '/var/empty/Library/Developer/Xcode/DerivedData/libtelnet-eparumqhzcujzafrpztuxqtzgfao/Logs/Build': (513) You don’t have permission to save the file “Build” in the folder “Logs”.}. User info: {
    NSLocalizedDescription = "Unable to create log store directory at '/var/empty/Library/Developer/Xcode/DerivedData/libtelnet-eparumqhzcujzafrpztuxqtzgfao/Logs/Build': (513) You don\U2019t have permission to save the file \U201cBuild\U201d in the folder \U201cLogs\U201d.";
}.`
`...`

- When building as root, the failure is clearly with the build system:
`cd ../remote_cmds-60/telnet.tproj`
`# Doesn't build with Fink's make`
`# Makefile includes recipes supplied as part of XCode that write and`
`# read /tmp/telnet unconditionally and without permission-checking (!)`
`/usr/bin/make \`
  `SRCROOT=../finkbuild/telnet/Sources \`
  `OBJROOT=../finkbuild/telnet/Build \`
  `SYMROOT=../finkbuild/telnet/Debug \`
  `DSTROOT=../finkbuild/telnet/Release`
`Installing source for telnet...`
`===== Building telnet RELEASE =====`
`/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc -arch x86_64 -g -Os -pipe -Wall -Werror -Wno-string-plus-int -fPIE -D__FBSDID=__RCSID -DTERMCAP -DKLUDGELINEMODE -DUSE_TERMIO -DENV_HACK -DAUTHENTICATION -DKRB5 -DSKEY -DIPSEC -DINET6 -DFORWARD -I. -I../finkbuild/telnet/Build/telnet -c -o ../finkbuild/telnet/Build/telnet/authenc.o authenc.c`
`authenc.c:39:10: fatal error: 'sys/cdefs.h' file not found`
`#include <sys/cdefs.h>`

**Comment by nieder** [comment]
Aquaterm needed to add `-UseNewBuildSystem=NO` to the xcodebuild command on Xcode10 and avoid similar /var/empty errors. Does that work for telnet?

**Comment by akhansen** [comment]
Unfortunately, it seems not to.

**Comment by jwhowarth** [comment]
You might try setting the environmentals for SDKROOT and MACOSX_DEPLOYMENT_TARGET which is what homebrew does.

[https://github.com/Homebrew/homebrew-core/blob/master/Formula/telnet.rb](url)

Regarding the reported permission errors in DerivedData, you might try passing a directory path in the fink build directory using the -derivedDataPath on the xcodebuild calls.

**Comment by akhansen** [comment]
Thanks.  Either of those gets me to a telnet executable.

However, the Makefile doesn't honor the variables that we'd been using for earlier Xcodes and so it builds everything in /tmp/telnet:

===== Linking telnet RELEASE =====
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc -arch x86_64  -lcurses -lkrb4 -lkrb5 -lipsec ../../libtelnet-13/build/Release/libtelnet.a          -dead_strip -pie -o /tmp/telnet/Debug/telnet \
		 /tmp/telnet/Build/telnet/authenc.o  /tmp/telnet/Build/telnet/commands.o  /tmp/telnet/Build/telnet/main.o  /tmp/telnet/Build/telnet/network.o  /tmp/telnet/Build/telnet/ring.o  /tmp/telnet/Build/telnet/sys_bsd.o  /tmp/telnet/Build/telnet/telnet.o  /tmp/telnet/Build/telnet/terminal.o  /tmp/telnet/Build/telnet/tn3270.o  /tmp/telnet/Build/telnet/utilities.o
clang: warning: argument unused during compilation: '-pie' [-Wunused-command-line-argument]
dsymutil --out /tmp/telnet/Debug/telnet.dSYM \
		/tmp/telnet/Debug/telnet || true
  SRCROOT=../finkbuild/telnet/Sources \
  OBJROOT=../finkbuild/telnet/Build \
  SYMROOT=../finkbuild/telnet/Debug \
  DSTROOT=../finkbuild/telnet/Release
/bin/rm -rf /sw/src/fink.build/root-telnet-60-1
/bin/mkdir -p /sw/src/fink.build/root-telnet-60-1/sw
/bin/mkdir -p /sw/src/fink.build/root-telnet-60-1/DEBIAN
/usr/sbin/chown -R fink-bld:fink-bld /sw/src/fink.build/root-telnet-60-1
sudo -u fink-bld [ENV] sh -c /tmp/fink.HEtnm
/tmp/fink.BLVlO
#!/bin/sh -ev
mkdir -p /sw/src/fink.build/root-telnet-60-1/sw/bin /sw/src/fink.build/root-telnet-60-1/sw/share/man/man1
install -m 555 ../finkbuild/telnet/Debug/telnet /sw/src/fink.build/root-telnet-60-1/sw/bin/telnet
install: ../finkbuild/telnet/Debug/telnet: No such file or directory

**Comment by jwhowarth** [comment]
You might have to restructure the build to match the Homebrew locations. Specifically, they are passing "SYMROOT=build" to the xcodebuild call and then...

OBJROOT=build/Intermediates SYMROOT=build/Products DSTROOT=build/Archive" 

to the make call as well as setting -isystembuild/Products/ on CFLAGS and -Lbuild/Products/ on LDFLAGS.

We will also need to convert the changes...

libtelnet_dst = buildpath/"telnet.tproj/build/Products"
libtelnet_dst.install "build/Release/libtelnet.a"
libtelnet_dst.install "build/Release/usr/local/include/libtelnet/"

bin.install "telnet.tproj/build/Archive/usr/local/bin/telnet"
man1.install "telnet.tproj/telnet.1"

into forms appropriate for our build system. Most importantly, we need to puzzle out how Homebrew positions the extracted remote_cmds-60.tar.gz vs the extracted libtelnet-13.tar.gz in their build system to match that or adjust for our differences in positioning.

**Comment by jwhowarth** [comment]
Looking at https://www.rubydoc.info/github/Homebrew/brew/Resource, it seems that they are decoupling the two builds by first building libtelnet and installing its files in build/Release which is then accessed with additional flags on CFLAGS and LDFLAGS for the make on telnet. I assume that there must be symlinks involved to relate the build/Release positioning of files to the -isystembuild/Products/ and -Lbuild/Products/ used in make.

**Comment by akhansen** [comment]
Or they just don't care about having intermediate stuff in /tmp/telnet.

As best as I can tell, that comes in as follows:

- _remote_cmds-60/Makefile_ has: `include $(MAKEFILEPATH)/CoreOS/ReleaseControl/BSDCommon.make`

- _/Applications/Xcode.app/Contents/Developer/Makefiles/CoreOS/ReleaseControl/BSDCommon.make_ has: `include $(MAKEFILEPATH)/CoreOS/ReleaseControl/Common.make`

- _/Applications/Xcode.app/Contents/Developer/Makefiles/CoreOS/ReleaseControl/Common.make_ has the following:
`##`
`# Some reasonable defaults for RC variables`
`##`
` `
`RC_ARCHS   = $(shell for i in `file /usr/lib/libSystem.B.dylib | grep 'shared library ' | sed 's|.*shared library ||'`; do $(CC) -arch $$i -E -x c /dev/null > /dev/null 2>&1 && echo $$i; done)`
`RC_RELEASE = unknown`
`RC_VERSION = unknown`
` `
`SRCROOT = /tmp/$(ProjectName)/Sources`
`OBJROOT = /tmp/$(ProjectName)/Build`
`SYMROOT = /tmp/$(ProjectName)/Debug`
`DSTROOT = /tmp/$(ProjectName)/Release`

I don't see a place in that file chain where those defaults can get replaced by user-specified values.

**Comment by jwhowarth** [comment]
What happens if you back up to your previous observations and just readjust the manual installation lines for telnet and telnet.1 to use their new locations in /tmp? As long as passing  -derivedDataPath suppresses the previous permission failures under fink-bld, I would think you are okay.

**Comment by akhansen** [comment]
No, I'm pretty sure it wasn't building in /tmp before, because the InstallScript would have failed to find the file before.  I've got a build from an earlier OS X/Xcode, so it definitely used to work.

**Comment by akhansen** [comment]
I've got something that "works" in #379, but I'm not thrilled with it.

**Comment by jwhowarth** [comment]
FYI, this issue is tracked at MacPorts at https://trac.macports.org/ticket/57137 with their current workaround being the same.

**Comment by akhansen** [comment]
Closed by 62a51543a12b8bcff4bb4f278da36d821519347e

**Comment by akhansen** [comment]
Reopening because Xcode 10.2 produces 
```
cd ../remote_cmds-60/telnet.tproj
# Doesn't build with Fink's make
# Makefile includes recipes supplied as part of XCode that write and
# read /tmp/telnet unconditionally and without permission-checking (!)
/usr/bin/make \
SRCROOT=../finkbuild/telnet/Sources \
OBJROOT=../finkbuild/telnet/Build \
SYMROOT=../finkbuild/telnet/Debug \
DSTROOT=../finkbuild/telnet/Release
Installing source for telnet...
===== Building telnet RELEASE =====
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc -arch x86_64 -g -Os -pipe -Wall -Werror -Wno-string-plus-int -fPIE -D__FBSDID=__RCSID -DTERMCAP -DKLUDGELINEMODE -DUSE_TERMIO -DENV_HACK -DAUTHENTICATION -DKRB5 -DSKEY -DIPSEC -DINET6 -DFORWARD -I. -I../finkbuild/telnet/Build/telnet -c -o ../finkbuild/telnet/Build/telnet/authenc.o authenc.c
authenc.c:39:10: fatal error: 'sys/cdefs.h' file not found
#include <sys/cdefs.h>
```
Again.

**Comment by akhansen** [comment]
ping

**Comment by nieder** [comment]
I tried adding -I$SDKPATH/usr/include after that leading -I. in PatchScript, but that led to a bunch of pointer/nullability errors in _stdio.h. I also tried to create symlinks to a sys directory inside the telnet build directory, but I stopped after 7 symlinks because it didn't look like it would stop anytime soon.

**Comment by nieder** [comment]
Fixed by 2b0e349 and 9324bd3, no?

**Comment by akhansen** [comment]
Oops, right.
